ipi/engine/thermostats.py

```diff
@@ -194,7 +194,104 @@ def step(self):
+         coupled thermostat.
+      sigma: Force noise amplitude sigma. Larger sigma means larger additional damping term.
+      sigtau: Sigma-Temperature coupling time constant tau.
+      T: Coefficient of the diffusive contribution of the thermostat, i.e. the
+         drift back towards equilibrium. Depends on tau and the time step.
+      S: Coefficient of the stochastic contribution of the thermostat, i.e.
+         the uncorrelated Gaussian noise. Depends on T and the temperature.
+   """
+
+   def get_T(self):
+      """Calculates the coefficient of the overall drift of the velocities."""
+
+      SigT = self.sigma**2/(Constants.kb*self.temp)
+
+      if self.tau > 0: LangT = np.exp(-self.dt/self.tau)
+      else: LangT = 1.0
+
+      NFLT = LangT - SigT
+      if NFLT < 0: NFLT = 0
+
+      return np.sqrt( NFLT )
+
+   def get_S(self):
+      """Calculates the coefficient of the white noise."""
+
+      if self.tau > 0:
+         return np.sqrt(Constants.kb*self.temp*(1.0 - np.exp(-self.dt/self.tau)))
+      else:
+         return 0.0
+
+   def __init__(self, temp = 1.0, dt = 1.0, tau = 0, sigma = 1.0, sigtau = 0, ethermo=0.0):
+      """Initialises ThermoNFL.
+
+      Args:
+         temp: The simulation temperature. Defaults to 1.0.
+         dt: The simulation time step. Defaults to 1.0.
+         tau: The thermostat damping timescale. Defaults to 0 (off).
+         sigma: Estimated force noise amplitude. Defaults to 1.0.
+         sigtau: Sigma-Temperature coupling time constant. Defaults to 0 (off).
+         ethermo: The initial heat energy transferred to the bath.
+            Defaults to 0.0. Will be non-zero if the thermostat is
+            initialised from a checkpoint file.
+      """
+
+      super(ThermoNFL,self).__init__(temp, dt, ethermo)
+
+      self.sigstep = False
+      dset(self,"tau",depend_value(value=tau,name='tau'))
+      dset(self,"sigma",depend_value(value=sigma,name='sigma'))
+      dset(self,"sigtau",depend_value(value=sigtau,name='sigtau'))
+      dset(self,"T",
+         depend_value(name="T",func=self.get_T,
+            dependencies=[dget(self,"temp"), dget(self,"tau"), dget(self,"dt"), dget(self,"sigma")]))
+      dset(self,"S",
+         depend_value(name="S",func=self.get_S,
+            dependencies=[dget(self,"temp"), dget(self,"tau"), dget(self,"dt")]))
+
+
+   def step(self):
+      """Updates the bound momentum vector with a langevin thermostat."""
+
+      et = self.ethermo
+      p = depstrip(self.p).copy()
+      sm = depstrip(self.sm)
+
+      p /= sm
+
+      et += np.dot(p,p)*0.5
+      p *= self.T
+      p += self.S*self.prng.gvec(len(p))
+      et -= np.dot(p,p)*0.5
+
+      p *= sm
+
+      self.p = p
+      self.ethermo = et
+
+      if self.sigtau > 0 and self.sigstep:
+         ekin = np.dot(depstrip(self.p),depstrip(self.p)/depstrip(self.m))*0.5
+         mytemp = ekin/Constants.kb/self.ndof * 2
+         self.sigma += (mytemp - self.temp) / self.sigtau * self.dt
```

Comment:
Also here, sigma should be a force, and you say that it equals a temperature divided time square.